src/dune_pkg/toolchain.ml

```diff
@@ -279,41 +287,46 @@ let install version =
 ;;
 
 let get ~log version =
-  let open Fiber.O in
-  let log_print style pp =
-    match log with
-    | `Never -> ()
-    | _ -> User_message.print (User_message.make [ Pp.tag style pp ])
-  in
-  if Version.is_installed version
-  then (
-    (match log with
-     | `Always ->
-       log_print Success
-       @@ Pp.textf
-            "Version %s of the compiler toolchain is already installed in %s"
-            (Version.to_string version)
-            (Version.target_dir version |> Path.Outside_build_dir.to_string)
-     | _ -> ());
-    Fiber.return ())
-  else (
-    let compiler_package = Compiler_package.of_version version in
-    log_print Details
-    @@ Pp.textf
-         "Will install version %s of the compiler toolchain to %s"
-         (Version.to_string version)
-         (Version.target_dir version |> Path.Outside_build_dir.to_string);
-    log_print Details @@ Pp.text "Downloading...";
-    let* () = fetch compiler_package in
-    log_print Details @@ Pp.text "Configuring...";
-    let* () = configure version in
-    log_print Details @@ Pp.text "Building...";
-    let* () = build version in
-    log_print Details @@ Pp.text "Installing...";
-    let+ () = install version in
-    log_print Success
-    @@ Pp.textf
-         "Success! Compiler toolchain version %s installed to %s."
-         (Version.to_string version)
-         (Version.target_dir version |> Path.Outside_build_dir.to_string))
+  Flock.with_flock
+    (Version.flock_path version)
+    ~name_for_messages:(sprintf "toolchain version %s" (Version.to_string version))
+    ~timeout_s:infinity
```

Comment:
While it is hard to know how long it should wait for another dune building the compiler, I do think that erroring out if the other instance gets stuck on something could be the better approach. Maybe it should not attempt to acquire the lock multiple times at and just print something along the lines of "well, some other instance (pid %d) is building the toolchain compiler".

WDYT?

Reply:
I'll definitely update this to print a message if acquiring the lock failed the first time around. I tried this earlier but gave up because it printed a bunch of seemingly spurious messages, but I just realized that it was because multiple fibers within the same instance of dune try to take the lock at the same time. I'll make it so that it only prints a message if the lock is currently taken by a different process.

Reply:
As for timing out, I'm happy to just let the user decide to kill it if it takes too long. I'm not sure about erroring out if another instance of dune is taking a long time to install the compiler. I don't want to make the experience worse for people with slow computers or slow internet where installing dune will take longer than expected. We could set a really conservative time limit like an hour, but my preference would be to make it clear to the user that this instance of dune is waiting for another instance of dune to finish building the compiler, and leave it up to them whether to kill the current instance of dune or wait for the other instance to finish.